Fixed #28010 - Added support for SELECT...FOR UPDATE OF

bluetech opened the pull request:
https://code.djangoproject.com/ticket/28010

Implements `select_for_update(of_self=True)`, meaning only rows of the base model are locked, rather than all selected rows.

This is a feature for Postgres and Oracle. I did not test on Oracle, since I do not have access to that. But I made sure that the query is good when `select_for_update_of_column = True` (as in the Oracle backend).

charettes commented:
Thanks for that!

I thought the plan was to allow something along the lines of `select_for_update(of=['self', 'referenced_table'])` to allow granular selection of tables instead?

charettes commented on django/db/models/query.py:
```diff
@@ -820,7 +820,7 @@ def intersection(self, *other_qs):
         return self._combinator_query('difference', *other_qs)
 
-    def select_for_update(self, nowait=False, skip_locked=False):
+    def select_for_update(self, nowait=False, skip_locked=False, of_self=False):
```
I thought `of` would be a list of paths to referenced tables with `self` being a special way to referenced the current table instead from the ticket description. So we could do something along the lines of:

e.g.
```python
Comment.objects.select_for_update(of={'author'})
```

Would result in
```
SELECT * FROM comment JOIN user ON (comment.user_id = user.id) FOR UPDATE OF user
```

bluetech commented:
I will try to implement the full `of=<fields>` support when I have time.

The reasons I didn't were:

- I think it would be a niche feature. `select_for_update` really feels like something for the "primary object" to me, something you would usually do at the beginning of the transaction with a fresh queryset. But maybe there are use cases for locking related objects. On the other hand `of={'self'}` should be common.

- I didn't have to dig too deep into the ORM code in order to implement just `of_self` (assuming it's correct). I tried to look at how `select_related` is implemented, since it seems like `of={...}` would have to do some similar things. But it would take some effort to understand what is going on there, and I'll probably miss many edge cases. I would also need to understand how the aliases are handled (pg wants a table alias, oracle wants an unaliased column name).

bluetech commented:
The tests failed because of a a conflict with 695d4dd790 which was just committed. I pushed a fix.

charettes commented:
While I agree full `of` support would necessitate a lot more work I don't think it would be a niche feature; excluding `self` would be while including it with another alias wouldn't (e.g. `of={'self', 'other'}` should be common enough).

My main concern with shipping partial support through this `of_self` API is that we'd have to consider deprecating it if we ever add complete support through `of`. The chosen name is also a bit misleading as it's really _of_self_only_.

bluetech commented:
Updated to support `of={'self', <related field>, ...}`. I don't know if this is the best approach but it seems to work well. There are some things that are not possible, for example specifying a table that is joined through `extra(tables=[...])` and similar, but I think it's OK.

BTW, if anyone knows of some way to test on Oracle, I'll be happy to try it.

orf commented on django/db/backends/base/operations.py:
```diff
@@ -177,16 +177,26 @@ def force_no_ordering(self):
         return []
 
-    def for_update_sql(self, nowait=False, skip_locked=False):
+    def for_update_sql(self, nowait=False, skip_locked=False, of=[]):
```
I never think mutable keyword arguments are a good idea, even in this case where they will never be mutated. I am surprised `flake8` didn't pick this up.

charettes replied:
@orf we don't have [flake8-mutable](https://pypi.python.org/pypi/flake8-mutable) installed on CI.

charettes commented:
Awesome work @bluetech! I use [this setup](https://code.djangoproject.com/wiki/OracleTestSetup) to run the tests against Oracle locally.

charettes commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

bluetech commented:
I fixed the Oracle test failure (hopefully).

I also changed `set()` to `()`. The reasons I preferred it over `None` are:

1. It provides a hint that a sequence is expected.
2. If I used `None`, it might imply that `()` is a non-special value, i.e. that *no* rows will be locked. But that's not possible.

felixxm commented on django/db/models/sql/query.py:
```diff
@@ -162,6 +162,7 @@ def __init__(self, model, where=WhereNode):
         self.select_for_update = False
         self.select_for_update_nowait = False
         self.select_for_update_skip_locked = False
+        self.select_for_update_of = {}
```
I think that `self.select_for_update_of` should be a tuple rather than a dictionary.

felixxm commented on docs/releases/2.0.txt:
```diff
@@ -131,7 +131,9 @@ CSRF
-* ...
+* Added the ``of`` argument to :meth:`.QuerySet.select_for_update()`,
+  supported on PostgreSQL and Oracle, to only lock rows from specific
+  models rather then all selected rows.
```
`then` -> `than`

felixxm commented on docs/ref/models/querysets.txt:
```diff
@@ -1635,14 +1635,23 @@ queryset is evaluated. You can also ignore locked rows by using
+:meth:`~django.db.models.query.QuerySet.select_related()` are also locked,
+in addition to rows of the base model. If this is not desirable, you can
+specify the specific related objects you want to lock in
+``select_for_update(of={...})``, similarly to
```
I think that we should be consistent with `of` type and use `tuple` everywhere (including docs, tests and examples).

felixxm commented on django/db/backends/base/operations.py:
```diff
@@ -177,16 +177,26 @@ def force_no_ordering(self):
+        else:
+            skip_locked_sql = ''
+
+        return 'FOR UPDATE%s%s%s' % (of_sql, nowait_sql, skip_locked_sql)
```
I think that:
```python
return 'FOR UPDATE%s%s%s' % (
    ' OF %s' % ', '.join(of) if of else '',
    ' NOWAIT' if nowait else '',
    ' SKIP LOCKED' if skip_locked else '',
)
```
is more readable.

charettes commented on django/db/models/sql/compiler.py:
```diff
@@ -467,14 +467,21 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
                         raise DatabaseError('SKIP LOCKED is not supported on this database backend.')
-                    for_update_part = self.connection.ops.for_update_sql(nowait=nowait, skip_locked=skip_locked)
+                    elif of and not self.connection.features.has_select_for_update_of:
+                        raise DatabaseError('FOR UPDATE OF is not supported on this database backend.')
```
Agreed but given all the other unsupported `select_for_update()` features raise `DatabaseError` as well this could be done in a subsequent PR.

felixxm replied:
You're right. I created a separate PR #8341.

charettes commented on django/db/models/sql/compiler.py:
```diff
@@ -802,6 +809,35 @@ def get_related_klass_infos(klass_info, related_klass_infos):
+        result = []
+
+        for name in self.query.select_for_update_of:
+            parts = [] if name == 'self' else name.split(LOOKUP_SEP)
```
I think there's a constant for `'self'` as well somewhere in `django.db`.

bluetech replied:
I found

```
django/db/models/fields/related.py:RECURSIVE_RELATIONSHIP_CONSTANT = 'self'
```

I don't think it's appropriate for this case. I can define a new constant if you think it will be useful. Or perhaps we can rename this constant to something more generic like `OWN_MODEL_CONSTANT`.

charettes commented on django/db/models/sql/compiler.py:
```diff
@@ -802,6 +809,35 @@ def get_related_klass_infos(klass_info, related_klass_infos):
+                    raise FieldError(
+                        "Bad field given in select_for_update(of={...}): '%s'. "
+                        "Only relational fields followed in the query are allowed." % name
+                    )
```
I wonder if we could enumerate valid options like we do for `select_related()`/`prefetch_related()`.

bluetech commented on django/db/models/sql/compiler.py:
```diff
@@ -802,6 +809,35 @@ def get_related_klass_infos(klass_info, related_klass_infos):
+        Return a quoted list of arguments to the SELECT FOR UPDATE OF
+        part of the query.
+        """
+        result = []
```
I used a list here because I build it up with `append`, and building up a tuple doesn't feel natural.

It also matches the code like in `get_distinct`: https://github.com/django/django/blob/5dbf1c4b23cda915369f4895be293369575238d0/django/db/models/sql/compiler.py#L580

bluetech commented:
@charettes 

> I was wondering how select_for_update() behaves in regard to fields that haven't been select_related() and what should be done in this case. Do you have any thoughts on that?

Fields that are not relations error out. For relation fields which are not `select_related` I saw two options:

1. Error out.
2. Make it induce a join using `setup_joins` (like I think `distinct` does).

I thought that 2 made no sense here, because we only want to limit locking of already-selected rows, not add new ones. Hence this case also raises an error. However I should make sure to test this case.

bluetech commented:
Updated:

- Rebased on top of #8346 (still not committed but wanted to preempt conflicts).
- Updated `for_update_sql` as suggested by @felixxm.
- Use `()` and tuples consistently.
- Added "Choices are: ..." to error message.
- Added a sentence to the changelog entry explaining when you might want to use `of`.
- Added a couple more test cases.

timgraham commented on django/db/models/sql/compiler.py:
```diff
@@ -802,6 +810,54 @@ def get_related_klass_infos(klass_info, related_klass_infos):
+                        klass_info = related_klass_info
+                        break
+                else:
+                    raise FieldError(
```
1. Did you try putting this validation in `django.db.models.query.QuerySet`. That's where I'd expect it to live, if feasible.
2. Is using the value of 'self' going to block fetching relations of the same name?

bluetech replied:
1. It would be very difficult to move this error to earlier, since I rely on the `klass_info` data structure.

2. If `'self'` is an actual valid name, then yes. The meaning of `'self'` in `select_for_update(of=(..., 'self', ...))` is always the "root" table.

timgraham replied:
I guess `self` probably can't be used as a field name very well, e.g.
```
>>> MyObject.objects.create(data='bar', self=my1)
TypeError: manager_method() got multiple values for argument 'self'
```

timgraham replied:
Is it possible to list all invalid fields rather than just one of them?

timgraham commented on docs/ref/databases.txt:
```diff
@@ -567,9 +567,9 @@ both MySQL and Django will attempt to convert the values from UTC to local time.
 -------------------------------------------------
 
-MySQL does not support the ``NOWAIT`` and ``SKIP LOCKED`` options to the
+MySQL does not support the ``NOWAIT``, ``SKIP LOCKED`` and ``OF`` options to the
```
Use the serial comma before "and", "or", etc. For example: A, B, and C

timgraham commented on docs/releases/2.0.txt:
```diff
@@ -131,7 +131,11 @@ CSRF
 ~~~~~~~~~~~~~~~~~
 
-* ...
+* Added the ``of`` argument to :meth:`.QuerySet.select_for_update()`,
```
I would put this under "Models".

bluetech commented:
I rebased and squashed and CI is green again.

timgraham commented on tests/select_for_update/tests.py:
```diff
@@ -152,6 +176,55 @@ def test_unsupported_skip_locked_raises_error(self):
+        """
+        for name in ['nonexistent', 'name', 'born__nonexistent', 'born__name']:
+            with self.subTest(name=name):
+                with self.assertRaisesMessage(
```
We generally use this indentation style for `assertRaisesMessage()`:
https://github.com/django/django/blob/3297dede7fce4b190f7b3bf0b0fc29a734151b61/tests/many_to_one/tests.py#L111-L116

timgraham commented on tests/select_for_update/tests.py:
```diff
@@ -182,7 +255,7 @@ def test_for_update_requires_transaction_only_in_execution(self):
     @skipUnlessDBFeature('supports_select_for_update_with_limit')
     def test_select_for_update_with_limit(self):
-        other = Person.objects.create(name='Grappeli')
+        other = Person.objects.create(name='Grappeli', born=self.city1, died=self.city2)
```
Maybe the fields be nullable to avoid the need for this change?

bluetech replied:
I've left it for now, because making the fields nullable runs into errors about the interaction of SELECT FOR UPDATE and outer joins.

I can try harder if necessary :)

timgraham commented on tests/select_for_update/tests.py:
```diff
@@ -322,3 +395,99 @@ def test_ordered_select_for_update(self):
+        object.
+        """
+        with transaction.atomic():
+            list(Person.objects.select_related(
```
Should we check the SQL or something here?

timgraham commented on tests/select_for_update/tests.py:
```diff
@@ -322,3 +395,99 @@ def test_ordered_select_for_update(self):
             qs = Person.objects.filter(id__in=Person.objects.order_by('-id').select_for_update())
             self.assertIn('ORDER BY', str(qs.query))
+
+    def run_select_for_update_with_related_objects_scenario(self, of):
```
These tests are a bit complicated -- do you think they offer substantial value? I didn't check the origin of the existing threading tests but I feel like it's testing the database more than Django. If Django generates the correct query, is there something else that could go wrong?

bluetech replied:
I agree, this is testing the DB, not Django. I removed them.

BTW, there are similar tests which already exist in the file for NOWAIT, that's what I cargo-culted in the first place.

timgraham commented on tests/select_for_update/tests.py:
```diff
@@ -322,3 +395,99 @@ def test_ordered_select_for_update(self):
+    @skipUnlessDBFeature('has_select_for_update_nowait')
+    def test_select_for_update_without_of_locks_all_objects(self):
+        """
+        select_for_update() with of=() should lock all rows
```
should lock -> locks

felixxm commented on django/db/models/sql/compiler.py:
```diff
@@ -832,6 +840,64 @@ def get_related_klass_infos(klass_info, related_klass_infos):
+            raise FieldError(
+                'Invalid field name(s) given in select_for_update(of=(...)): %s. '
+                'Only relational fields followed in the query are allowed. '
+                'Choices are: %s' % (
```
Dot is missing: `'Choices are: %s.'`.
